src/audio/copier/copier_dai.c

```diff
@@ -450,40 +449,81 @@ static int copy_single_channel_c32(const struct audio_stream *src,
 	return 0;
 }
 
+static void copier_dai_adjust_params(const struct copier_data *cd,
+				     struct ipc4_audio_format *in_fmt,
+				     struct ipc4_audio_format *out_fmt)
+{
+	struct comp_buffer *dma_buf;
+	int dma_buf_channels;
+	int dma_buf_container_bits, dma_buf_valid_bits;
+
+	/* Call this func only for DAI gateway with already setup DMA buffer */
+	assert(cd->dd[0] && cd->dd[0]->dma_buffer);
+	dma_buf = cd->dd[0]->dma_buffer;
+
+	/* Unfortunately, there is a mess regarding how the gateway DMA buffer format
+	 * is set up. The number of channels can come from hardware parameters (extracted
+	 * from a blob?) and also appears in the copier's input/output format. In case
+	 * the value returned by the hardware looks valid, it should take precedence over
+	 * the value from the copier's input/output format.
+	 *
+	 * The frame format comes from the topology as dev->ipc_config.frame_fmt and also
+	 * comes as the copier's input/output format. The logic is confusing: the format
+	 * from the topology takes priority, except when the copier's format container and
+	 * valid sample size are different. Perhaps this is to support the 16-bit valid
+	 * in the 32-bit container format used by SSP, as such a format cannot be specified
+	 * in the topology?
+	 */
+	dma_buf_channels = audio_stream_get_channels(&dma_buf->stream);
+	dma_buf_container_bits = audio_stream_sample_bytes(&dma_buf->stream) * 8;
+	dma_buf_valid_bits = get_sample_bitdepth(audio_stream_get_frm_fmt(&dma_buf->stream));
+
+	if (cd->direction == SOF_IPC_STREAM_PLAYBACK) {
+		out_fmt->channels_count = dma_buf_channels;
+
+		if (!(dma_buf_container_bits == out_fmt->depth &&
+		      out_fmt->depth != out_fmt->valid_bit_depth)) {
```

Comment:
is this because `out_fmt->valid_bit_depth` hasn't been set up yet or is it because it must be changed? So what we previously thought was a valid bit depth, now it's a different one? Something is a bit strange here. Your `if` is the same as
```
if (dma_buf_container_bits != out_fmt->depth || out_fmt->depth == out_fmt->valid_bit_depth)
```
So we can enter this branch also if `dma_buf_container_bits == out_fmt->depth` and then you set `out_fmt->depth = dma_buf_container_bits` again... Could you double-check this?

Reply:
The setup of DMA buffer format and conversoin func is somewhat confusing. I wrote a big comment above this code trying to explain. The second part of the comment is about this weird logic.

DMA buffer format is set in dai-zephyr from dev->ipc_config.frame_fmt, which comes from topology. That is enum sof_ipc_frame type like SOF_IPC_FRAME_S16_LE, SOF_IPC_FRAME_S32_LE, etc. There is no way to use this value to setup DMA buffer, for example, for 16-bit valid in 32-bit container -- no such enum value. Such format is used in our CI for SSP gateway.
So before my PR we had this code to fix the problem (see line 466 in copier_dai.c in version before my PR):
```
/*
		 * dai_zephyr_params assigns the conversion function
		 * based on the input/output formats but does not take
		 * the valid bits into account. So change the conversion
		 * function if the valid bits are different from the
		 * container size.
		 */
		audio_stream_fmt_conversion(in_fmt->depth,
					    in_fmt->valid_bit_depth,
					    &in_bits, &in_valid_bits,
					    in_fmt->s_type);
		audio_stream_fmt_conversion(out_fmt->depth,
					    out_fmt->valid_bit_depth,
					    &out_bits, &out_valid_bits,
					    out_fmt->s_type);

		if (in_bits != in_valid_bits || out_bits != out_valid_bits)
			cd->dd[0]->process =
				cd->converter[IPC4_COPIER_GATEWAY_PIN];
```
The code overwrites cd->dd[0]->process previously set in dai-zephyr based on dev->ipc_config.frame_fmt and other pin format with new value based on coper input and output formats in case of ``` if (in_bits != in_valid_bits || out_bits != out_valid_bits) ```.
That condition looks strange to me but we lived with it for quite a sometime.

The condition in my PR (also looking very confusing) works like this:
* if dma_buf_container_bits != in_fmt->depth, then in_fmt values cannot be trusted and are ignored (format from topology takes precedence over copier's input/output formats).
* if dma_buf_container_bits == in_fmt->depth, we still prefer value from topology except for in_fmt->depth != in_fmt->valid_bit_depth. This is to handle that 16-bit valid in 32-bit container SSP case. Topology cannot specify such format, so we have to rely on copier's input/output format in this case.

Of course, it would be much simpler if we just always use copier's input/output formats -- they contain all the required info. However, on our CI we have test that actually sends different container sizes for topology format and copier input/output format. So that weird logic is required.